### PR TITLE
feat: add integration tests for repo setup automation

### LINKED ISSUE #17: Add integration tests for repo_setup.sh
## Description

Create comprehensive integration tests for the repository setup automation, including the new CONTRIBUTING.md template generation feature.

## Background

The `repo_setup.sh` script now:
- Creates new repositories with language-specific templates (Java, C, C++, Rust)
- Adds artagon-common and artagon-license as submodules
- Configures git hooks and agent preferences
- **NEW**: Generates CONTRIBUTING.md from template with GitHub variables (PR #16)

We need integration tests to ensure this automation works correctly across all supported languages and configurations.

### Test Coverage Needed

1. **Basic Repository Creation**
   - [ ] Java project creation with Maven
   - [ ] C project creation with CMake
   - [ ] C project creation with Bazel
   - [ ] C++ project creation with CMake
   - [ ] C++ project creation with Bazel
   - [ ] Rust project creation with Cargo

2. **CONTRIBUTING.md Generation**
   - [ ] Template file exists and is accessible
   - [ ] Variable substitution works (repository.name, repository.owner, repository.description)
   - [ ] Special characters in descriptions are properly escaped
   - [ ] Generated file has correct markdown structure
   - [ ] Links to artagon-common resources are valid

3. **Submodule Integration**
   - [ ] artagon-common submodule initialized correctly
   - [ ] artagon-license submodule initialized correctly
   - [ ] License files exported properly
   - [ ] Nix submodule added when --with-nix flag used

4. **Agent Configuration**
   - [ ] Codex/Claude agent directories created
   - [ ] Agent preferences synced from artagon-common
   - [ ] Symlinks created correctly

5. **Git Configuration**
   - [ ] Hooks path set to artagon-common/git-hooks
   - [ ] Initial commit created with proper message format
   - [ ] Repository pushed to GitHub successfully

6. **Error Handling**
   - [ ] Invalid project types rejected
   - [ ] Missing required parameters detected
   - [ ] Authentication failures handled gracefully
   - [ ] Cleanup on failure

### Test Framework
- Use Bash testing framework (bats-core or similar)
- Create isolated test environments (temp directories)
- Mock GitHub CLI operations where possible
- Clean up test repositories after completion

### Test Structure
```bash
tests/integration/
├── test_repo_setup.sh           # Main test suite
├── test_contributing_gen.sh     # CONTRIBUTING.md specific tests
├── fixtures/
│   ├── test-descriptions.txt    # Edge case descriptions
│   └── expected-outputs/        # Expected file contents
└── helpers/
    └── test-helpers.sh          # Shared test utilities
```

### Example Test Cases

```bash
# Test CONTRIBUTING.md generation
test_contributing_md_basic() {
  local test_dir=$(mktemp -d)
  cd $test_dir
  
  # Mock repo_setup.sh execution
  PROJECT_NAME="test-project"
  OWNER="artagon"
  DESCRIPTION="Test project for integration tests"
  
  # Run CONTRIBUTING.md generation
  .common/artagon-common/scripts/gh_setup_contributing.sh \
    --repo-name "$PROJECT_NAME" \
    --repo-owner "$OWNER" \
    --repo-desc "$DESCRIPTION" \
    --force
  
  # Verify output
  assert_file_exists "CONTRIBUTING.md"
  assert_file_contains "CONTRIBUTING.md" "test-project"
  assert_file_contains "CONTRIBUTING.md" "artagon"
  assert_file_contains "CONTRIBUTING.md" "Test project for integration tests"
  
  # Cleanup
  rm -rf $test_dir
}

# Test special character escaping
test_contributing_md_special_chars() {
  local desc="Project with & pipe | slash / backslash \\"
  
  # Should properly escape sed special characters
  # Should not cause sed errors
  # Should preserve the characters in output
}
```

## Acceptance Criteria

- [ ] Integration test suite created in `tests/integration/`
- [ ] All 6 test categories implemented
- [ ] Tests run in CI/CD pipeline
- [ ] Tests pass for all supported project types
- [ ] Edge cases covered (special characters, empty descriptions, etc.)
- [ ] Test documentation added to TESTING.md
- [ ] Mock mode available for local testing without GitHub API

## Related

- PR #16: CONTRIBUTING.md template feature
- Issue #15: CONTRIBUTING.md template creation
- `scripts/repo_setup.sh`: Main repository setup script
- `scripts/gh_setup_contributing.sh`: CONTRIBUTING.md generator

## Estimate

**Effort**: Medium (2-3 days)
**Priority**: High (ensures automation reliability)

### PR DESCRIPTION
**Pull request by trumpyla**
## Summary

Implements comprehensive integration testing framework for repository setup automation scripts, addressing #17.

### Test Infrastructure
- **Test Helpers** (`tests/integration/helpers/test-helpers.sh`)
  - Assertion functions: `assert_file_exists`, `assert_file_contains`, `assert_file_not_contains`
  - Test environment management: `create_test_env`, `cleanup_test_env`
  - Test counter tracking with summary reporting

### Test Suites

- **CONTRIBUTING.md Generation Tests** (`test_contributing_gen.sh`)
  - 7 comprehensive test cases covering:
    - Basic variable substitution
    - Special character escaping (& | / \\)
    - Empty description handling
    - Idempotency verification
    - Parameter validation
    - Git repository auto-detection
    - Template variable replacement validation

- **Basic Validation Tests** (`test_basic_validation.sh`)
  - Template file existence checks
  - Script executability verification
  - Syntax validation
  - Template variable presence
  - Integration point validation

- **Repository Setup Tests** (`test_repo_setup_simple.sh`)
  - Component validation for repo_setup.sh
  - Help output functionality
  - Parameter validation
  - Invalid input rejection
  - CONTRIBUTING.md integration verification

- **Master Test Runner** (`run_all_tests.sh`)
  - Quick validation of 8 critical components
  - Inline tests for fast execution
  - Clear pass/fail reporting

### Documentation

- **Updated TESTING.md** (lines 138-415)
  - Complete integration testing guide
  - Test structure documentation
  - Running instructions for all test suites
  - Edge case testing examples
  - Writing new tests guide with template
  - CI/CD integration notes
  - Known limitations documented

## Test Coverage

The integration tests cover:
- ✅ Template file validation
- ✅ Variable substitution and escaping
- ✅ Script syntax and executability
- ✅ Parameter validation and error handling
- ✅ Git auto-detection functionality
- ✅ Idempotency guarantees
- ✅ Integration points between scripts
- ✅ Edge cases (special characters, empty values)

## Testing

```bash
# Run all integration tests
./tests/integration/run_all_tests.sh

# Run specific test suites
./tests/integration/test_contributing_gen.sh
./tests/integration/test_basic_validation.sh
./tests/integration/test_repo_setup_simple.sh
```

## Notes

- Test framework uses isolated temporary directories for each test
- All tests clean up after themselves
- Tests are designed to run without requiring GitHub API access
- Full end-to-end repository creation tests require manual execution with GitHub authentication

Closes #17

### CODE REVIEW

File: tests/integration/test_contributing_gen.sh [hunk 1]
```diff
@@ -0,0 +1,265 @@
...
+  git init -q
+  git config user.email "test@test.com"
+  git config user.name "Test"
+
+  mkdir -p .common/artagon-common
+  cp -r "$ROOT_DIR/templates" .common/artagon-common/
+  cp -r "$ROOT_DIR/scripts" .common/artagon-common/
+
+  # Test with special sed characters: & | / \
+  local desc="Project with & ampersand | pipe / slash \\ backslash"
```
[the remark this answers is not in the document]
Comment (trumpyla) [reply tests/integration/test_contributing_gen.sh]: Fixed in commit 996e1a0 (now 5ecba12 after rebase). Changed to use `printf` with single quotes to ensure a literal backslash is included in the test string, and added assertion to verify 'backslash' appears in the output.

File: tests/integration/test_contributing_gen.sh [hunk 2]
```diff
@@ -0,0 +1,268 @@
...
+  git config user.name "Test"
+
+  mkdir -p .common/artagon-common
+  cp -r "$ROOT_DIR/templates" .common/artagon-common/
+  cp -r "$ROOT_DIR/scripts" .common/artagon-common/
+
+  # Test with special sed characters: & | / and backslash
+  # Use printf to ensure literal backslash is included
+  local desc
+  desc=$(printf 'Project with & ampersand | pipe / slash \\ backslash')
```
[the remark this answers is not in the document]
Comment (trumpyla) [reply tests/integration/test_contributing_gen.sh]: Fixed in commit d961ad7. Added assertions to check for the actual special characters (&, |, /, \\) in the output, not just the descriptive words. This ensures proper escaping and preservation of special sed characters through the template substitution process.

File: tests/integration/helpers/test-helpers.sh [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+# Run command and capture exit code safely
+# Usage: run_and_capture_exit <command> [args...]
+# Returns: exit code of the command
+run_and_capture_exit() {
+  set +e
+  "$@"
+  local exit_code=$?
+  set -e
+  return $exit_code
+}
```
[the remark this answers is not in the document]
Comment (trumpyla) [reply tests/integration/helpers/test-helpers.sh]: Fixed in commit 5d65b6c. Renamed function from `run_and_capture_exit` to `run_without_errexit` to accurately reflect its purpose. Updated usage pattern to call the function directly in conditionals (`if run_without_errexit ...; then`) which properly handles both success and failure cases with set -e enabled. Enhanced documentation to clarify usage.
